src/fmu/dataio/providers/_fmu.py

```diff
@@ -84,318 +97,239 @@ class FmuProvider:
+            logger.debug("Using proposed casepath %s", self.casepath_proposed)
+            return self.casepath_proposed
 
-            self._iter_path = Path("/" + "/".join(folders))
-            self._real_path = Path("/" + "/".join(folders[:-1]))
+        logger.debug("Try detect casepath from RUNPATH %s", self._runpath)
+        if self._runpath:
+            if _casepath_has_metadata(self._runpath.parents[1]):
+                return self._runpath.parents[1]
 
-            self._iter_id = int(str(FmuEnv.ITERATION_NUMBER.value))
-            self._real_id = int(str(FmuEnv.REALIZATION_NUMBER.value))
+            if _casepath_has_metadata(self._runpath.parent):
+                return self._runpath.parent
 
-    def _read_case_metadata(self) -> None:
-        """Check if metadatafile file for CASE exists, and if so parse metadata.
-
-        If file does not exist, still give a proposed file path, but the
-        self.casepath_proposed_metadata will be {} (empty) and the physical file
-        will not be made.
-        """
-        logger.debug("Read case metadata, if any...")
-        if not self._casepath:
-            logger.info("No case path detected, hence FMU metadata will be empty.")
-            return
-
-        case_metafile = Path(self._casepath) / ERT_RELATIVE_CASE_METADATA_FILE
-        if case_metafile.exists():
-            logger.debug("Case metadata file exists in file %s", str(case_metafile))
-            self._case_metadata = ut.yaml_load(case_metafile, loader="standard")
-            logger.debug("Case metadata are: %s", self._case_metadata)
-        else:
-            logger.debug("Case metadata file does not exists as %s", str(case_metafile))
-            warn(
-                "Cannot read case metadata, hence stop retrieving FMU data!",
-                UserWarning,
+        if self.fmu_context == FmuContext.CASE:
+            raise ValueError(
+                "Could not auto detect the casepath, please provide it as input."
             )
-            self._case_metadata = {}
-
-    def _read_optional_restart_data(self) -> None:
-        # Load restart_from information
-        logger.debug("Read optional restart data, if any, and requested...")
-        if not self._case_metadata:
-            return
 
-        if not environ.get(RESTART_PATH_ENVNAME):
-            return
+        logger.debug("No case metadata, issue a warning!")
+        warn("Case metadata does not exist, metadata will be empty!", UserWarning)
+        return None
 
+    def _get_restart_data_uuid(self) -> UUID | None:
+        """Load restart_from information"""
+        assert self._runpath is not None
         logger.debug("Detected a restart run from environment variable")
-        restart_path = Path(self._iter_path) / environ[RESTART_PATH_ENVNAME]
-        restart_iter = self._get_folderlist_from_path(restart_path)[-1]
+        restart_path = self._runpath / environ[RESTART_PATH_ENVNAME]
```

Comment:
Is there possibly missing exception handling here? The previous code had
```python
if not environ.get(RESTART_PATH_ENVNAME):
    return
```
prior to this dict access. If something like that is no longer needed, it's still preferable to use `.get` or `os.getenv` I think